Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h

```diff
@@ -0,0 +1,429 @@
+                if (!found_edge || decimation_props.decimation_cost < m_decimation_map[optimal_edge].decimation_cost)
+                {
+                    optimal_edge = edge;
+                    found_edge = true;
+                }
+            }
+
+            // if no valid edges then stop
+            if (!found_edge) break;
+
+#ifdef CGAL_APPROX_DECOMPOSITION_VERBOSE            
+//            std::cout << std::endl;
+            std::cout << "#" << num_vertices(m_graph) << " Optimal edge for decimation: " << optimal_edge << std::endl;
+            std::cout << "Decimation cost: " << m_decimation_map[optimal_edge].decimation_cost << ", Concavity value: " << m_decimation_map[optimal_edge].new_cluster_props.concavity << std::endl;
+            std::cout << "Total edges: " << num_edges(m_graph) << std::endl;
+//            BOOST_FOREACH(graph_edge_descriptor edge, edges(m_graph))
+//            {
+//                std::cout << source(edge, m_graph) << " " << target(edge, m_graph) << std::endl;
+//            }
+#endif
+
+            // decimate optimal valid edge and update the decimation costs of modified edges
+            decimate_edge(optimal_edge, concavity_threshold, CONCAVITY_FACTOR);
+        }
+
+        // resulting number of produced clusters
+        std::size_t num_clusters = num_vertices(m_graph);    
+
+        // current cluster's id
+        int cluster_id = 0;
+
+        BOOST_FOREACH(graph_vertex_descriptor vert, vertices(m_graph))
+        {
+            Cluster_properties& cluster_props = m_cluster_map[vert];
+
+            // assign id to all current cluster's faces
+            BOOST_FOREACH(face_descriptor face, cluster_props.faces)
+            {
+                face_ids[face] = cluster_id;
+            }
+
+            ++cluster_id;
+        }
+
+        // post check if all faces are assigned to any cluster
+        BOOST_FOREACH(face_descriptor face, faces(m_mesh))
+        {
+            CGAL_assertion(face_ids[face] >= 0 && face_ids[face] < num_clusters);
+        }
+
+        return num_clusters;
+    }
+
+private:
+    const TriangleMesh& m_mesh;
+    const GeomTraits& m_traits;
+    
+    Graph m_graph; // adjacency list
+
+    Graph_cluster_map m_cluster_map; // vertex property map of the adjacency list
+    Graph_decimation_map m_decimation_map; // edge property map of the adjacency list
+
+//    std::priority_queue<graph_edge_descriptor> m_candidates;
+
+    Concavity<TriangleMesh, GeomTraits> m_concavity_calc; // concavity calculator that computes concavity values for any subset of faces of the input mesh
+
+    template <class Mesh>
+    struct Noborder_predicate
+    {
+        Noborder_predicate(const Mesh& mesh) : m_mesh(mesh) {}
+        
+        bool operator()(const edge_descriptor& edge) const { return !CGAL::is_border(edge, m_mesh); }
+        
+        const Mesh& m_mesh;
+    };
+
+    struct Cluster_properties
```

Comment:
Comment me: _struct used when evaluating the weight of collapsing an edge...._

Reply:
Exactly. It contains all the necessary information about a cluster. While computing the weight (decimation cost) of edge, information about the merged cluster is computed and is stored in the same struct inside Edge_properties.

Reply:
I mean add comment in the code to avoid reverse engineering